internal/core/src/mmap/ChunkedColumnGroup.h

```diff
@@ -0,0 +1,322 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+
+#include <folly/io/IOBuf.h>
+#include <sys/mman.h>
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <memory>
+#include <vector>
+#include <math.h>
+
+#include "cachinglayer/CacheSlot.h"
+#include "cachinglayer/Manager.h"
+#include "cachinglayer/Translator.h"
+#include "cachinglayer/Utils.h"
+
+#include "common/Array.h"
+#include "common/Chunk.h"
+#include "common/GroupChunk.h"
+#include "common/Common.h"
+#include "common/EasyAssert.h"
+#include "common/Span.h"
+#include "common/Array.h"
+#include "mmap/ChunkedColumn.h"
+#include "mmap/ChunkedColumnInterface.h"
+
+namespace milvus {
+
+using GroupChunkVector = std::vector<std::shared_ptr<GroupChunk>>;
+
+using namespace milvus::cachinglayer;
+
+// ChunkedColumnGroup represents a collection of group chunks
+class ChunkedColumnGroup {
+ public:
+    explicit ChunkedColumnGroup(
+        std::unique_ptr<Translator<GroupChunk>> translator)
+        : slot_(Manager::GetInstance().CreateCacheSlot(std::move(translator))) {
+    }
+
+    virtual ~ChunkedColumnGroup() = default;
+
+    // Get the number of group chunks
+    size_t
+    num_chunks() const {
+        return slot_->num_cells();
+    }
+
+    PinWrapper<GroupChunk*>
+    GetGroupChunk(int64_t chunk_id) const {
+        auto ca = SemiInlineGet(slot_->PinCells({chunk_id}));
+        auto chunk = ca->get_cell_of(chunk_id);
+        return PinWrapper<GroupChunk*>(ca, chunk);
+    }
+
+    int64_t
+    NumRows() const {
+        int64_t total_rows = 0;
```

Comment:
we definitely want to cache "number of rows in each chunk", it should be metadata that is prefetched and cached, do not compute metadata based on data. even if this chunk is already loaded, this incurs locking in cache.